### ISSUE TITLE
false positive: `no-unreachable` inside `catch` block when throwing value created outside scope of `try`

### ISSUE THREAD

**Issue by jsejcksn** [denoland/deno_lint#1303]
### Lint Name

[`no-unreachable`](https://lint.deno.land/rules/no-unreachable)

### Code Snippet

`lint_fail_test.ts`:
```ts
import { assert } from "jsr:@std/assert@^1.0.1/assert";
import { assertThrows } from "jsr:@std/assert@^1.0.1/throws";

Deno.test("lint", () => {
  const error = assertThrows(
    () => {
      const fooError = new Error("foo");
      try {
        throw fooError;
      } catch (cause) {
        assert(cause instanceof Error && cause.message === "foo");
     // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        throw new Error("bar", { cause }); /*                    │
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                       │
        │                                                        │
        └─ error[no-unreachable]: This statement is unreachable ─┘ */
      }
    },
    Error,
    "bar",
  );

  assert(error.cause instanceof Error && error.cause.message === "foo");
});
```

### Expected Result

Running `deno lint lint_fail_test.ts` exits with status code `0`

### Actual Result

Exits with status code `1` (see additional info below)

### Additional Info

For comparison, here's the same test module except that the declaration of `fooError` is inside the `try` block:

`lint_pass_test.ts`:
```ts
import { assert } from "jsr:@std/assert@^1.0.1/assert";
import { assertThrows } from "jsr:@std/assert@^1.0.1/throws";

Deno.test("lint", () => {
  const error = assertThrows(
    () => {
      try {
        const fooError = new Error("foo");
        throw fooError;
      } catch (cause) {
        assert(cause instanceof Error && cause.message === "foo");
        throw new Error("bar", { cause });
      }
    },
    Error,
    "bar",
  );

  assert(error.cause instanceof Error && error.cause.message === "foo");
});
```

The module `lint_pass_test.ts` passes linting, and both test modules complete successfully using Deno's test runner. Here is terminal output for both modules:

```txt
% deno lint lint_pass_test.ts
Checked 1 file

% deno lint lint_fail_test.ts
error[no-unreachable]: This statement is unreachable
  --> /Users/deno/lint_fail_test.ts:11:9
   | 
11 |         assert(cause instanceof Error && cause.message === "foo");
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  docs: https://lint.deno.land/rules/no-unreachable


error[no-unreachable]: This statement is unreachable
  --> /Users/deno/lint_fail_test.ts:13:9
   | 
13 |         throw new Error("bar", { cause }); /*                    │
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  docs: https://lint.deno.land/rules/no-unreachable


Found 2 problems
Checked 1 file

% deno test *_test.ts
Check file:///Users/deno/lint_fail_test.ts
Check file:///Users/deno/lint_pass_test.ts
running 1 test from ./lint_fail_test.ts
lint ... ok (0ms)
running 1 test from ./lint_pass_test.ts
lint ... ok (0ms)

ok | 2 passed | 0 failed (10ms)
```

This was discovered in a CI failure here:
- https://github.com/denoland/std/actions/runs/10131143544/job/28013460563?pr=5561

…and might be a regression. Ref:
- https://github.com/denoland/deno_lint/issues/716

### Version

```
deno 1.45.4 (release, aarch64-apple-darwin)
v8 12.7.224.13
typescript 5.5.2
```

**Comment by bartlomieju** [comment]
Fixed in https://github.com/denoland/deno_lint/pull/1497
